Feature/limelight

oconnelc opened the pull request:
Making a pull request to review. I'll probably make some comments/make some changes.

anseljohn committed:
Add PID controller

anseljohn committed:
Update imports

anseljohn committed:
Start 2019 drive implementation

anseljohn committed:
Conversion from degree double to Rotation2d object

anseljohn committed:
Change class name

anseljohn committed:
Update TurnTo to use PIDController class

anseljohn committed:
PIDController getError

anseljohn committed:
Change setpoint and input configuration

- set contiuous for continuous input (circular)
- set input range for -180 to 180 degrees to calc shortest path
- change error calculation to use rotateby

anseljohn committed:
Merge remote-tracking branch 'origin/master' into experimental/turn-to

anseljohn committed:
Update constants and rotation2d location

GlebBabahov committed:
Merge remote-tracking branch 'origin/master' into feature/Limelight

anseljohn committed:
Merge branch 'master' into experimental/turn-to

GlebBabahov committed:
Added limelight wrapper class

GlebBabahov committed:
Added limelight wrapper class

anseljohn committed:
Change package

Control package removed

anseljohn committed:
Remove for PID pull request

GlebBabahov committed:
Merge remote-tracking branch 'origin/experimental/turn-to' into feature/Limelight

kc5tjn committed:
Merge remote-tracking branch 'origin/feature/Limelight' into feature/Limelight

kc5tjn committed:
Added Parameters and functions to support LimeLight target tracking 

I added parameters to SystemSettings.java for the calculating the distance to target and angle of approach to the target using the vision tracking values presented by the LimeLight camera.  The parameters are calibrated to the 2018 robot, sitting on a table to simulate a high mounted, offset camera.

I added a VisionTarget enum to hold values related to the specific target types.

I added methods to the LimeLight class to do the calculations, they are:
public double calcTargetDistance( double targetHeight )
public double calcTargetDistance( SystemSettings.GameTarget target )
public double calcTargetApproachAngle()
public Point calcTargetLocation( SystemSettings.VisionTarget target )

I also added a LimeLight.Point class to hold the resulting x,y coordinate from the calcTargetLocation function.  This should be replaced by an appropriate point class if one already exists in our system.  Should I have just used java.awt.Point?  I wasn't sure if we should pull in the java.awt packages.

These compile, but have not been tested on a system yet.  I can setup some junit tests if we're using junit.

I also had some questions about error handling in our system.

oconnelc committed:
Changed the Point class to use Poof's Translation2d

flybotix commented on common/src/main/java/us/ilite/common/config/SystemSettings.java:
```diff
@@ -104,4 +104,60 @@ public String getFileLocation() {
+    // TODO These values are specific to the targets, not the camera, and may belong elsewhere
+    // The current target values assume the limelight processing stream is configured to target
+    // the bottom of the vision target
+    public enum VisionTarget {
```
Possibly add one for the line on the ground (height = 0), and cargo ball (height = ... 0, or 6.5, or 13?).

Also consider adding the LL pipeline names to this enum.

flybotix commented on robot/src/main/java/us/ilite/robot/modules/Limelight.java:
```diff
@@ -0,0 +1,250 @@
+    }
+
+    
+
```
Should use ```Codex<ETargetingData>```, where ```enum ETargetingData implements CodexOf<Double>```.

This allows us to graph these values using existing shuffleboard stuff, including timestamps, row id's, etc.

oconnelc replied:
I'm not seeing any reference to ETargetingData in Limelight. Where in the code do you want this change?

flybotix commented on robot/src/main/java/us/ilite/robot/modules/Limelight.java:
```diff
@@ -0,0 +1,250 @@
+        return approachAngle;
+    }
+
+    public class Point {
```
This is fine for now.  We can refactor later to use the translate/rotate stuff from 254 once we start testing with the drivetrain.

oconnelc replied:
I'm using the Poof's Translation2D class.

flybotix commented on robot/src/main/java/us/ilite/robot/modules/Limelight.java:
```diff
@@ -0,0 +1,250 @@
+    public class TargetingData {
+        public final boolean tv;
+        public final double tx, ty, ta, ts, tl, tshort, tlong, thoriz, tvert;
+
```
For the Targetting Data itself, it may be beneficial to make this more generic than just the limelight as part of a 2nd module.  For example, we would make an interface that returned tx, tv, ty, ta, (etc), and the Limelight class would implement that interface.  In the implementation, the limelight class would pull from NT.  Then the 2nd module would use the LL data (via the generic interface) to do the calculations.

This would allow us to use additional cameras with opencv, multiple LL's, a Pixy, a Zed (one day...), etc.

oconnelc committed:
Added unit test for limelight

oconnelc committed:
Added additional targets

GlebBabahov committed:
Added target searching and locking

GlebBabahov committed:
Merge remote-tracking branch 'origin/feature/Limelight' into feature/Limelight

oconnelc committed:
Refactored the TargetingData

Created a new interface for targeting data with a specific implementation
for the Limelight. The location of these files may not be correct; we
should discuss where we should put the interface and implementation.

oconnelc committed:
Merge branch 'feature/Limelight' of github.com:iliterobotics/FRC-Robot-2019 into feature/Limelight

oconnelc committed:
Merge branch 'master' into feature/Limelight

oconnelc committed:
Fixed compile errors

oconnelc commented:
I don't know what's going on with my pushes. The code works in my local workspace, but then something happens in Github and then when I pull down, my workspace is all messed up. I'm working on it now.

oconnelc committed:
Fixed compile error caused by bad merge

oconnelc committed:
Took out unnecessary NoOp class and replaced with a lambda implementation

oconnelc committed:
Forgot to take out the import from the last commit

oconnelc committed:
Took out unit test since Limelight would need major refactoring

flybotix committed:
Refactored with Codexes
